Fixes #2943 - implements shared grpc connection per multiple EDS subscriptions.

dmitri-d opened the pull request:
Was originally started in https://github.com/envoyproxy/envoy/pull/4823. 

*Description*: Fixes #2943 by implementing shared grpc connection per multiple EDS subscriptions. 
*Risk Level*:
*Testing*:
*Docs Changes*:
*Release Notes*:
[Optional Fixes #Issue]
[Optional *Deprecated*:]

@htuch: I removed ClusterManager dependency on GrpcMuxFactory, as there was a circular dependency after some recent changed. The extacted code now lives in EdsSubscriptionFactory, and all changes I made are constrained within EDS.

dmitri-d commented:
* fixed broken tests

htuch commented on source/common/upstream/eds_subscription_factory.cc:
```diff
@@ -0,0 +1,50 @@
+
+    if (muxes_.find(mux_key) == muxes_.end()) {
+      muxes_.insert({mux_key,
+                     std::unique_ptr<Config::GrpcMuxImpl>(new Config::GrpcMuxImpl(local_info, std::move(async_client), dispatcher,
```
Nit; std::make_unique is preferred in C++14. Do you need to std::move?

htuch commented on source/common/upstream/eds_subscription_factory.cc:
```diff
@@ -12,8 +12,8 @@ namespace Upstream {
       muxes_.insert({mux_key,
-                     std::make_unique<Config::GrpcMuxImpl>(local_info, std::move(async_client), dispatcher,
-                                                           service_method, random, scope, rate_limit_settings)});
+                     std::move(std::make_unique<Config::GrpcMuxImpl>(local_info, std::move(async_client), dispatcher,
```
Maybe use `emplace`?

dmitri-d commented:
@htuch: looks like using emplace instead of insert did the trick.

htuch commented on source/common/upstream/eds_subscription_factory.h:
```diff
@@ -0,0 +1,35 @@
+  public:
+    EdsSubscriptionFactory() : muxes_() {}
+
+    Config::GrpcMux& getOrCreateMux(const LocalInfo::LocalInfo& local_info, Grpc::AsyncClientPtr async_client,
```
Should this be private? I don't think this is part of the interface.

dmitri-d replied:
It's not part of the interface, but I have tests for it. The only thing I can think of in order to keep tests (which I would like to do) is to subclass ```EdsSubscriptionFactory``` in tests and add a proxy method for ```getOrCreateMux```. Would that work?

htuch replied:
Yeah, you can either use a testing peer pattern or provide a public `getOrCreateMuxForTests` that makes clear this is only to be used in tests (and make the existing implementation private and the target).

htuch commented on source/common/upstream/eds_subscription_factory.cc:
```diff
@@ -0,0 +1,50 @@
+      return result;
+    }
+
+    auto to_return = Config::SubscriptionFactory::subscriptionFromConfigSource<
```
Just return directly?

dmitri-d replied:
fixed.

htuch commented on source/common/config/eds_grpc_subscription_impl.h:
```diff
@@ -0,0 +1,40 @@
+namespace Envoy {
+namespace Config {
+
+class EdsGrpcSubscriptionImpl : public Config::Subscription<envoy::api::v2::ClusterLoadAssignment> {
```
I'm still a bit confused what this is providing over `GrpcSubscriptionImpl`? They seem to have the same implementation.

dmitri-d replied:
Meant to ask you about this and then forgot. My understanding is that the instance of ```GrpcMux``` in ```GrpcSubscriptionImpl``` will be deallocated together with the main class. I need to pass a reference to ```GrpcMux``` to ```EdsGrpcSubscriptionImpl```, as the mux instances are managed by ```EdsSubscriptionFactory```. Subscription classes are very much the same, but I wasn't sure how to reconcile mux ownership...

htuch replied:
Sure, I think we can refactor this a bit, have a base `GrpcSubscriptionImpl` that takes a `GrpcMux&` and then an `OwnedGrpcSubscriptionImpl` subclass that creates and owns the `GrpcMux` for those use cases.

htuch commented on test/common/upstream/eds_subscription_factory_test.cc:
```diff
@@ -0,0 +1,71 @@
+  ASSERT_TRUE(&first_mux == &second_mux);
+}
+
+TEST_F(EdsSubscriptionFactoryTest, ShouldReturnDiffrentMuxes) {
```
Nit: s/Diffrent/Different/

dmitri-d replied:
fixed.

htuch commented on test/common/upstream/BUILD:
```diff
@@ -82,6 +82,19 @@ envoy_cc_test(
+        "//test/mocks/local_info:local_info_mocks",
+        "//test/mocks/config:config_mocks",
+        "//test/mocks/runtime:runtime_mocks",
+	"//test/mocks/stats:stats_mocks",
```
This probably needs buildifier run on it (fix_format).

htuch commented on test/common/upstream/eds_subscription_factory_test.cc:
```diff
@@ -0,0 +1,108 @@
+  ASSERT_TRUE(&first_mux != &second_mux);
+}
+
+TEST_F(EdsSubscriptionFactoryTest, ShouldUseGetOrCreateMuxWhenApiConfigSourceIsUsed) {
```
Can you add a single line `//` above each test explaining what it does in natural language?

dmitri-d replied:
done.

htuch commented on test/common/upstream/eds_subscription_factory_test.cc:
```diff
@@ -0,0 +1,108 @@
+                                                     random_, config.api_config_source(), stats_store_, rate_limit_settings_);
+
+  Config::Subscription<envoy::api::v2::ClusterLoadAssignment>& eds_subscription = *subscription;
+  ASSERT_TRUE(&expected_mux == &(dynamic_cast<Config::EdsGrpcSubscriptionImpl&>(eds_subscription).grpcMux()));
```
I think you are after `EXPECT` in all these final test statements, and specifically `EXPECT_EQ` here.

dmitri-d replied:
done.

dmitri-d commented:
* rebased
* renamed EdsGrpcSubscriptionImpl to GrpcManagedMuxSubscriptionImpl
* Refactored GrpcSubscriptionImpl to become a wrapper around GrpcManagedMuxSubscptionImpl
* Reduced visibility of EdsSubscriptionFactory::getOrCreateMux to protected

Still need to fix formatting.

htuch commented on source/common/upstream/eds_subscription_factory.cc:
```diff
@@ -0,0 +1,57 @@
+    Runtime::RandomGenerator& random, const ::envoy::api::v2::core::ApiConfigSource& config_source,
+    Stats::Scope& scope, const Config::RateLimitSettings& rate_limit_settings) {
+
+  const uint64_t mux_key = MessageUtil::hash(config_source.grpc_services(0));
```
Nit: remove blank lines above/beyond, Google C++ style encourages dense packing of lines except where needed to logically divide up sections.

htuch commented on source/common/upstream/eds_subscription_factory.cc:
```diff
@@ -0,0 +1,57 @@
+    std::function<Config::Subscription<envoy::api::v2::ClusterLoadAssignment>*()>
+        rest_legacy_constructor,
+    const std::string& rest_method, const std::string& grpc_method) {
+
```
Nit: same here.

htuch commented on source/common/config/grpc_managed_mux_subscription_impl.h:
```diff
@@ -0,0 +1,41 @@
+namespace Config {
+
+template <class ResourceType>
+class GrpcManagedMuxSubscriptionImpl : public Config::Subscription<ResourceType> {
```
We should make https://github.com/envoyproxy/envoy/blob/master/source/common/config/grpc_subscription_impl.h a subclass of this as discussed on Slack.

htuch commented on source/common/upstream/eds_subscription_factory.h:
```diff
@@ -0,0 +1,41 @@
+
+namespace Envoy {
+namespace Upstream {
+class EdsSubscriptionFactory {
```
A comment explaining what this does would be helpful, motivation, link back to original issue. Also a TODO explaining that it would make sense to generalize this for RDS in the future.

dmitri-d commented:
* removed empty spaces as pointed above
* added a reference to the original issue + a TODO note

dmitri-d commented:
* signed-off and rebased

dmitri-d commented:
* fixed formatting and clang_tidy issues

I'll squash and sign-off the commit once the latest changes will have been ok'd.

dmitri-d commented:
/retest

dmitri-d commented:
* rebased

dmitri-d commented:
/restest

dmitri-d commented:
Thanks for the reviews and the merge, @htuch.
